### DISCUSSION TITLE
Consider adding support of some streaming server [like Audiobookshelf].

### DISCUSSION THREAD

**Discussion by zkvsky** [PaulWoitaschek/Voice#2385] [Ideas]
Probably the most commonly used and fully featured right now is https://github.com/advplyr/audiobookshelf

Audiobookshelf in particular is amazing for organizing and pulling metadata.

**Comment by zkvsky** [comment]
And here's the API https://api.audiobookshelf.org/#introduction

**Reply by asislife20** [reply]
Absolutely would be a great feature! I use emby for streaming , but its crap for audiobooks. I was able to add audiobooks to this app from an smb server using cifs documents provider. It has an option to used smb as local storage, and works fairly well!

**Comment by hkfuertes** [comment]
That would be awesome, there are no native good apps for ABS, and this app is awesome. My setup now is having both, ABS and Voice, and have a shared folder between them, but it would be really nice if we can save the progress online.

This being said, I thought that we could have a special "directory" appear when logged in, for the books in the server, and an extra button on the book player to download/remove it. UI-wise, that should not imply much more... in settings a button to open the Oauth/weblogin so that the app stores the token... and on every "onload" a request to the server if its up to refresh the token.

Also on the downloaded folder/book, maybe a metadata file indicating that it is indeed an online book, that the server is X, the token is Y, the id is Z, so regularly, or upon app close the timestamp can be syncronized.

I'm a developer, but never tried kotlin at this level... I will try to play with this idea... but maybe if there is interest, a collaborator can make this changes quicker than me (or even... just do them at all hahahahah), but I guess this is a very reasonable foundation on how to implement it 🤷‍♂️ 

This is just for audiobooks (podcast... will need a rethink... could be a v2)

**Reply by hkfuertes** [reply]
Even simpler... in the initial release, v0 let us (the users) create this metadata file manually:
```jsonc
// /sdcard/books/<author>/<title>/abs_metadata.json
{
   "host": "http://192.168.77.10:13378",
   "user": <user>,
   "password": <password>,
   "book_id": 1
}
```

**Comment by hkfuertes** [comment]
One other option... I see that there is some Tasker integration already... maybe triggering a tasker task with the timestamp on play/pause...

**Comment by sandreas** [comment]
> That would be awesome, there are no native good apps for ABS

@hkfuertes 
Normally I wouldn't suggest another project in a projects discussion, but there is [Lissen](https://github.com/GrakovNe/lissen-android/). I don't think that it is anywhere near Voice in terms of code quality or playing ability, but it is a native android player for ABS :-)

**Comment by sandreas** [comment]
@zkvsky 
@hkfuertes 
@asislife20 

So recently I started a little experiment where the main Voice data entry points for retrieving audio book information are relocated in a central `VoiceMediaSource`. The idea is to have an interface with a very limited set of API methods (e.g. `query`, `bookmarks`, `storeBookPosition`, etc.) which is totally abstract and not bound to the local storage of files nor the scanning or media analysis.

Here is the branch: https://github.com/PaulWoitaschek/Voice/compare/main...mediasource-api

Around this we would need a managing `VoiceMediaSourceRepo` where we could `list` the available media sources, then `add`,  `delete` and `change` - as well as a `VoiceMediaSourceService` where we can `activate` or `switch` the current media source if there is more than one.

This is A LOT of work, but it might be worth the time to support more or less ANY media source that supports the required APIs.
To start somewhere, we could implement a `LocalMediaSource` (for locally stored books) but also an `AudioBookShelfMediaSource`, a `JellyfinMediaSource` etc. with only the effort of having to implement the one specific HTTP Api and a configuration UI instead of having to manually integrate everything. 

I still need to clarify with @PaulWoitaschek, whether this might out of scope for Voice in general (or maybe just too huge).

**Reply by zkvsky** [reply]
great work! I will look into it, eventually... (I'm swamped with project atm)

**Reply by entropicdrifter** [reply]
Just a random passerby but it seems to me that this is a fantastic way to leverage a modular architecture like Voice has.

**Comment by MikotoMisaka10032** [comment]
I would also love to see Audiobookshelf support in Voice.  \
It would be a fantastic addition.

**Comment by sandreas** [comment]
@MikotoMisaka10032 
Meanwhile, you could use the Audiobookshelf App to "manage" your local books by downloading them into a shared folder and then use Voice to listen. That's what I do ATM.

@PaulWoitaschek
What do you think about the generic MediaSource API (or how we'd call it, because MediaSource is already used by google... maybe MediaStore)? Is this something we'd like to chase or do we focus on local playback instead?

To see my very rough draft from long ago, you could take a look at  this: https://github.com/PaulWoitaschek/Voice/compare/main...mediasource-api
